README.md
# GeneidX

GeneidX provides a fast annotation of the protein-coding genes in an eukaryotic genome taking as input the genome assembly and the taxonomic ID of the species to annotate.

In the description here, you can find our preliminary results, a schema of our method and a description of the minimal requirements and commands required for running it.

Stay tuned for an article with detailed descriptions and feel free to [contact us](mailto:ferriol.calvet@crg.eu) if you are trying it and find any problem.


## Preliminary results
 The results of an initial benchmarking using vertebrates genomes annotated in Ensembl show that our method is **as accurate as the top *ab initio* gene predictors**, but it is **between 10 and 100 times faster**.
<!-- ![Summary of vertebrate benchmark](images/Benchmarking4GitHubX.svg) -->



## Running GeneidX
Having defined the parameters and ensuring that Nextflow and a container technology are installed.

This pipeline requires the compressed genome assembly and the taxid of the species to annotate.
```
export NXF_VER=22.04.4
nextflow run guigolab/geneidx -profile <docker/singularity>
                                        --genome <GENOME>.fa.gz
                                        --taxid <TAXID>
                                        --outdir <OUTPUT_directory>
```

or alternatively, clone the repository and then run it (highly recommended)
```
git clone https://github.com/guigolab/geneidx.git
cd geneidx
export NXF_VER=22.04.4
nextflow run main.nf -profile <docker/singularity>
                      --genome <GENOME>.fa.gz
                      --taxid <TAXID>
                      --outdir <OUTPUT_directory>
```


Revise the DETAILS section below for the minor specifications of each parameter.

Geneidx is working in Nextflow version 22.04.4 although we are working on updating the code up to make it work with the latest versions.


## Before running GeneidX
1. **Make sure ( Docker or Singularity ) and Nextflow are installed in your computer.**
  - [Docker](https://docs.docker.com/engine/install/)
  - [Singularity](https://sylabs.io/guides/3.0/user-guide/installation.html#)
  - [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html#installation)

2. Define the input parameters:
  - Compressed FASTA file with `.fa.gz` termination.
  - Taxid of the species to annotate or from the closest relative described with taxid. You can find this information in [NCBI Taxonomy database](https://www.ncbi.nlm.nih.gov/taxonomy).


3. Optional parameters:
  - Proteins from closely related species: (choose one of these options)
      - Provide a compressed FASTA file with the proteins selected.
      - Let the pipeline download the proteins automatically from UniRef90 (nothing should be provided in this case.)
  - Tune parameters for the gene prediction process.
      - Indicate the values for each Geneid parameter.
      - Let the pipeline select them from the closest pretrained Geneid parameter file. (find the parameters in data/Parameter_files.taxid/)
      - Find more information here: [GENEID parameter files](https://genome.crg.es/software/geneid/index.html#parameters)



## Schema
Which steps are taking place as part of this pipeline?
This is a graphical summary, and the specific steps are outlined below.
![Summary of vertebrate benchmark](images/SchemaWhite.png)
1. Get the set of proteins to be used for the protein-to-genome alignments.
2. Get the closest Geneid parameter file to use, as source of the parameters is not indicated by the user.
3. Create the protein database for DIAMOND to use it as a source.
4. Align the provided genome against the database created using DIAMOND BLASTx flavour.
5. Run the auto-training process:
  - Use matches to estimate the coding sections, look for open reading frames between stop codons.
  - Use matches from the same protein to predict the potential introns.
  - From the sequences of both previous steps compute the initial and transition probability matrices required for the computation of the coding potential of the genome that will be annotated.
6. Update the parameter file with the parameters indicated in the *params.config* file and also the matrices automatically generated from the protein-DNA matches.
7. Run Geneid with the new parameter file and the protein-DNA matches from the previous steps as additional evidence.
This is done in parallel for each independent sequence inside the genome FASTA file, and it consists of the following steps:
  - Pre-process the matches to join those that overlap and re-score them appropriately.
  - Run Geneid using the evidence and obtain the GFF3 file.
  - Remove the files from the internal steps.
8. Concatenate all the outputs into a single GFF3 file that is sorted by coordinates.
9. Add information from proteins matching the predicted genes.


## DETAILS:
  - **The name of the sequences in the FASTA file cannot contain unusual characters.**
  - **The input genome file must be a gzip-compressed FASTA file. (.fa.gz)**
  - **Auto-train the parameter file always.**

  - It is recommended to clone the repository and then run the pipeline from there.
  - So far the output of the predictions is not stored in the path indicated when running the pipeline, but in output/species/{taxid of the species}.
  - If you are running the pipeline multiple times, it is recommended that you define a directory for downloading the docker/singularity images to avoid having to download them multiple times. See `singularity.cacheDir variable` in `nextflow.config`.
  - If you are starting from an unmasked genome assembly, check another version of this pipeline integrating the genome masking process. [GeneidXmask](https://github.com/FerriolCalvet/geneidXmask) repository.

  - If you have used Geneid in the past and have manually trained a parameter file, we are open to receive them and share them in our repositories giving credit to the users who generated them.
  Contact us at [ferriol.calvet@crg.eu](mailto:ferriol.calvet@crg.eu).


Follow us on Twitter ([@GuigoLab](https://twitter.com/GuigoLab)) for updates in the article describing our method, and contact us for any questions or suggestions.


.nextflow/history
2022-09-05 16:35:47	2m 43s	scruffy_avogadro	OK	bfb3b7a0f6a6bee6fd2f518d34978e89	a62b782b-3258-48bb-b6fe-c6f687118e48	nextflow run main.nf -with-docker
2022-09-05 17:53:26	1m 22s	amazing_hawking	ERR	cc9b29e6c4646333c1e6919ca40927c6	865aaf3a-cb10-442e-830c-bc2408075ad2	nextflow run main.nf -with-docker
2022-09-05 17:55:41	18.7s	curious_volta	OK	dcbe4272e9b3772298d87e01d799a52f	865aaf3a-cb10-442e-830c-bc2408075ad2	nextflow run main.nf -with-docker -resume
2022-09-05 18:06:46	1m 49s	sick_poisson	OK	22c88729f458a3e5a2a74e2b47cb563e	22b20c92-2fe2-4cc6-bb89-b902131a7964	nextflow run main.nf -with-docker
2022-09-10 16:09:02	4m 31s	determined_bernard	ERR	22c88729f458a3e5a2a74e2b47cb563e	14e46aec-7083-4d2a-9e45-5470b8da9cd0	nextflow run main.nf -with-docker
2022-09-10 16:14:08	14.9s	loving_austin	ERR	22c88729f458a3e5a2a74e2b47cb563e	45a42826-5bf7-45e1-9050-110e41a9913d	nextflow run main.nf -with-docker --prot-file data/SampleProtein.fa.gz
2022-09-10 16:14:29	34.8s	scruffy_keller	OK	22c88729f458a3e5a2a74e2b47cb563e	0f2bde47-a815-4e75-8c47-3bf7f2e8cdcd	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:31:18	-	chaotic_gutenberg	-	22c88729f458a3e5a2a74e2b47cb563e	8cd5b405-1071-4b1c-a7bf-83ee504c9657	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:31:40	34.4s	confident_kare	OK	e80ab3a84d11e535da7fb675c3f9cf64	9131c871-b948-4c35-8252-40756d1c9245	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:36:07	6.8s	stupefied_volta	ERR	e80ab3a84d11e535da7fb675c3f9cf64	0d011a2a-d894-4020-b8b9-da9f1a86d4d9	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:37:07	6.2s	mighty_pesquet	ERR	e80ab3a84d11e535da7fb675c3f9cf64	4c2513c3-94f5-4f46-9838-5d93171c28c6	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:37:53	6.1s	confident_booth	ERR	cc4df475c1fad587dd101a2731738545	2666c059-6ac5-4b29-84cb-ba1f451518ec	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:45:14	-	nostalgic_liskov	-	dd6ff2436a65c4d3485d9ecc6ba99608	91fa0d14-fdcc-474e-b2dd-632ff0bfa27e	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:45:33	6.3s	jovial_dalembert	ERR	e36f7ec1af38df743db0435c3ba7664f	984741c5-89db-4d2d-ab4d-97aefb43c99d	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:46:32	10.8s	nasty_celsius	ERR	e36f7ec1af38df743db0435c3ba7664f	3c5a0c3f-9ceb-466e-887b-50f6be2b05a4	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:48:06	33.3s	magical_thompson	OK	872dcf0466fba5e70b6f294f1d685ebb	5407c63d-fd44-4753-b720-ece1e55f1746	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:53:08	6.9s	soggy_chandrasekhar	ERR	872dcf0466fba5e70b6f294f1d685ebb	fb6298db-f73f-4b8a-b4dd-de19084ae508	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 16:54:25	34.1s	thirsty_kare	OK	872dcf0466fba5e70b6f294f1d685ebb	10df269d-f99d-4189-b6ff-ed1f2222f9c0	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:04:02	6.7s	amazing_visvesvaraya	ERR	872dcf0466fba5e70b6f294f1d685ebb	f023e3db-7051-4273-8400-61e8b7151d64	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:04:37	7.3s	drunk_poisson	ERR	872dcf0466fba5e70b6f294f1d685ebb	11938c46-7f78-4574-8a38-2c82ae543fed	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:05:12	33.7s	pensive_swirles	OK	872dcf0466fba5e70b6f294f1d685ebb	a9903be8-9b18-4570-a39a-ab8c1e3bc915	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:10:37	18.5s	infallible_liskov	ERR	4600a817ce32129732a3df8fc6a7285f	74b03eed-8a53-47cf-8ee9-b6c86b327942	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:12:28	33.1s	trusting_hoover	OK	4600a817ce32129732a3df8fc6a7285f	2743ce6e-e1c8-49d5-8b6b-fc020bd08bc8	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:15:23	34.1s	elated_majorana	OK	4600a817ce32129732a3df8fc6a7285f	1eff095b-872a-4756-b380-0f6e05c12704	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-10 17:18:25	33.5s	distracted_majorana	OK	4600a817ce32129732a3df8fc6a7285f	346023c7-d120-4434-8b3f-141bd431cd5a	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-11 10:23:16	34.4s	small_brown	OK	4600a817ce32129732a3df8fc6a7285f	3f50fa70-7236-4ff0-a048-1bc7ea12e45c	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-11 10:43:17	5s	condescending_wescoff	ERR	d6b068a0ac7a30e6e8aac1b35453d2b2	a48f5c5e-51e7-48f9-89b0-9ee7bc5e6641	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-11 10:43:58	34.7s	mighty_plateau	OK	d6b068a0ac7a30e6e8aac1b35453d2b2	ea3db8ac-98c1-45fc-954a-2ed66477e156	nextflow run main.nf -with-docker --prot_file data/SampleProtein.fa.gz
2022-09-11 10:45:49	3m 46s	voluminous_panini	OK	d6b068a0ac7a30e6e8aac1b35453d2b2	76891657-3280-4f0b-b189-64860c3257a3	nextflow run main.nf -with-docker
2022-09-13 23:38:49	8.3s	admiring_boyd	ERR	d6b068a0ac7a30e6e8aac1b35453d2b2	c815be51-29f0-43e4-9d84-b40016a44710	nextflow run main.nf -with-docker
2022-09-13 23:39:33	1m 21s	shrivelled_davinci	OK	d6b068a0ac7a30e6e8aac1b35453d2b2	4a1a5a40-b506-4750-a56d-d63d2dd009c3	nextflow run main.nf -with-docker
2022-09-16 15:30:34	-	elated_curran	-	9fbf7370168fa52805b39d0c443916a7	ef2fe779-9585-460e-8506-34d021811855	nextflow run main.nf -with-docker
2022-09-16 15:30:59	2m 39s	intergalactic_gilbert	ERR	b197e3bbe34a07d45eaba2203bcc3663	a0f52e3f-d9c9-49fb-afd1-94256cfb09c6	nextflow run main.nf -with-docker
2022-09-16 16:16:31	45.1s	distraught_meucci	OK	69998da09ec91202abafeb17e5bdef31	3ccfc8a9-1cb6-49dd-aa1f-5468bb9e905c	nextflow run main.nf -with-docker

scripts/mergeProteinLabels.py
#!/usr/bin/env python
# coding: utf-8

# In[12]:


import pandas as pd
import sys


# In[13]:


data = pd.read_table(sys.argv[1],
                     names=["seq", "source", "feature", "start", "end", "strand",
                           "score", "phase", "attributes", "protein"],
                     header = None)
data.head()


# In[16]:


grouped_data = data.groupby( ["seq", "source", "feature", "start", "end",
                              "strand", "score", "phase",
                              "attributes"]).agg({"protein" : ','.join }).reset_index()


# In[17]:


#grouped_data.head()


# In[21]:


updated_attributes = [ x if y == '.' else x + f";ProteinIDs={y}" for x, y in zip(grouped_data["attributes"].values, grouped_data["protein"].values) ]


# In[24]:


grouped_data["attributes"] = updated_attributes
grouped_data.drop("protein", axis = 1, inplace = True)



grouped_data.sort_values(["seq", "start", "end"],
              ascending = [True, True, False], inplace = True)
# In[25]:


grouped_data.to_csv(sys.argv[2],
                       sep = "\t",
                       header = None,
                       index = None)


# In[ ]:
